Mathlib/Analysis/Calculus/FirstDerivativeTest.lean

```diff
@@ -0,0 +1,145 @@
+
+
+/-- The First-Derivative Test from calculus, minima version. -/
+lemma localMin_of_deriv_Ioo {f : ℝ → ℝ} {a b c : ℝ} (h : ContinuousAt f b)
+    (g₀ : a < b) (g₁ : b < c)
+    (hd₀ : DifferentiableOn ℝ f (Set.Ioo a b)) (hd₁ : DifferentiableOn ℝ f (Set.Ioo b c))
+    (h₀ : ∀ x ∈ Set.Ioo a b, deriv f x ≤ 0)
+    (h₁ : ∀ x ∈ Set.Ioo b c, 0 ≤ deriv f x) : IsLocalMin f b := by
+    have := localMax_of_deriv_Ioo (f := -f) g₀ g₁
+      (by simp_all) (DifferentiableOn.neg hd₀) (DifferentiableOn.neg hd₁)
+      (fun x hx => deriv.neg (f := f) ▸ Left.nonneg_neg_iff.mpr <|h₀ x hx)
+      (fun x hx => deriv.neg (f := f) ▸ Left.neg_nonpos_iff.mpr <|h₁ x hx)
+    exact (neg_neg f) ▸ IsLocalMax.neg this
+
+/-- If `p` holds to the left of `a` then it holds in an open interval `(l, a)`. -/
+lemma Filter.Eventually.exists_lt_forall_Ioo {α : Type*} [TopologicalSpace α] [LinearOrder α]
+    [OrderTopology α] [NoMinOrder α] {a : α} {p : α → Prop} (h : ∀ᶠ x in 𝓝[<] a, p x) :
+    ∃ l < a, ∀ x ∈ Ioo l a, p x :=
+  mem_nhdsWithin_Iio_iff_exists_Ioo_subset.1 h
+
+/-- If `p` holds to the right of `a` then it holds in an open interval `(a, l)`. -/
+lemma Filter.Eventually.exists_gt_forall_Ioo {α : Type*} [TopologicalSpace α] [LinearOrder α]
+    [OrderTopology α] [NoMaxOrder α] {a : α} {p : α → Prop} (h : ∀ᶠ x in 𝓝[>] a, p x) :
+    ∃ l > a, ∀ x ∈ Ioo a l, p x :=
+  mem_nhdsWithin_Ioi_iff_exists_Ioo_subset.1 h
+
+
+/-- Monotonicity of open intervals under removal of `max` at a left endpoint. -/
+theorem mem_Ioo_of_mem_Ioo_max_left {R : Type*} [LinearOrder R]
+    {b u₀ v₀ x : R} (hx : x ∈ Ioo (max u₀ v₀) b) :
+    x ∈ Ioo u₀ b := by simp_all
+
+/-- Monotonicity of open intervals under removal of `min` at a right endpoint. -/
+theorem mem_Ioo_of_mem_Ioo_min_right {R : Type*} [LinearOrder R]
+    {b u₁ v₁ x : R} (hx : x ∈ Ioo b (min u₁ v₁)) :
+    x ∈ Ioo b u₁ := by simp_all
+
+/-- The interval inclusion `(a,b] \ {b} ⊆ (a,b)`. -/
+theorem mem_Ioo_of_mem_Ioc_of_ne {R : Type*} [LinearOrder R]
+    {b u₀ x : R} (hx : x ∈ Ioc u₀ b) (H : ¬x = b) :
+    x ∈ Ioo u₀ b := by
+    have := hx.2
+    simp_all only [mem_Ioc, and_true, mem_Ioo, true_and, gt_iff_lt]
+    exact lt_of_le_of_ne this H
+
+/-- The interval inclusion `[a,b) \ {a} ⊆ (a,b)`. -/
+theorem mem_Ioo_of_mem_Ico_of_ne {R : Type*} [LinearOrder R]
+    {b u₀ x : R} (hx : x ∈ Ico u₀ b) (H : ¬x = u₀) :
+    x ∈ Ioo u₀ b := by
+    have := hx.2
+    simp_all only [mem_Ico, and_true, mem_Ioo, gt_iff_lt]
+    exact lt_of_le_of_ne hx fun a ↦ H (id (Eq.symm a))
+
+
+ /-- The First-Derivative Test from calculus, maxima version,
+ expressed in terms of left and right filters. -/
+lemma localMax_of_deriv' {f : ℝ → ℝ} {b : ℝ} (h : ContinuousAt f b)
+    (hd₀ : ∀ᶠ x in 𝓝[<] b, DifferentiableAt ℝ f x) (hd₁ : ∀ᶠ x in 𝓝[>] b, DifferentiableAt ℝ f x)
+    (h₀  : ∀ᶠ x in 𝓝[<] b, 0 ≤ deriv f x) (h₁  : ∀ᶠ x in 𝓝[>] b, deriv f x ≤ 0) :
+    IsLocalMax f b := by
```

Comment:
I think this result more easily derivable from `localMax_of_deriv_Ioo`. Also, use `Filter.Eventually.and` first to conclude `∀ᶠ x in 𝓝[<] b, DifferentiableAt ℝ f x ∧ 0 ≤ deriv f x` and simplify the argument.

Reply:
I'm not sure how, do you mean something like use `rw [Filter.Eventually, mem_nhdsWithin] at hd₀` and then go from there to get `Ioo`?

Also, if I may ask: there is a merge conflict now apparently. What's the best method for resolving that?

Reply:
Oh, that's very nice and it removes the need for some of the lemmas. So I've removed them in the new version.